lib/e2e/src/error.rs

```diff
@@ -1,23 +1,69 @@
-use alloy::{contract::Error, sol_types::SolError};
+use alloy::sol_types::SolError;
 
 pub trait ErrorExt<E> {
-    /// Checks that `Self` corresponds to the typed abi-encoded error
+    /// checks that `self` corresponds to the typed abi-encoded error
     /// `expected`.
-    fn is(&self, expected: E) -> bool;
+    fn is_err(&self, expected: E) -> bool;
 }
 
-impl<E: SolError> ErrorExt<E> for Error {
-    fn is(&self, expected: E) -> bool {
+pub trait RevertExt {
+    /// checks that `self` corresponds to expected revert `code`.
+    /// E.g. for arithmetic overflow.
+    fn reverts(&self, code: i64) -> bool;
+}
+
+impl RevertExt for alloy::contract::Error {
+    fn reverts(&self, code: i64) -> bool {
+        let Self::TransportError(e) = self else {
+            return false;
+        };
+
+        //  ErrorResp(
+        //      ErrorPayload {
+        //          code: -32000,
+        //          message: "execution reverted",
+        //          data: None,
+        //      },
+        //  )
+        let payload = e.as_error_resp().expect("should contain payload");
+        payload.code == code
+    }
+}
+
+impl<E: SolError> ErrorExt<E> for alloy::contract::Error {
+    fn is_err(&self, expected: E) -> bool {
         let Self::TransportError(e) = self else {
             return false;
         };
 
         let raw_value = e
             .as_error_resp()
             .and_then(|payload| payload.data.clone())
-            .expect("should extract the error");
+            .unwrap_or_else(|| panic!("{:#?}", e));
+        // .expect("should extract the error");
         let actual = &raw_value.get().trim_matches('"')[2..];
         let expected = alloy::hex::encode(expected.abi_encode());
-        return expected == actual;
+        expected == actual
+    }
+}
+
+impl<E: SolError> ErrorExt<E> for eyre::ErrReport {
```

Comment:
@alexfertel @qalisander any better idea for this kind of error is welcome.

Reply:
Yeah, I think we don't need this one here. Since it's an `ErrReport` from `eyre`, we should be able to traverse the sources until we reach the Transport error.